Added  PHP 8.2 support, dropped PHP 7 support

glo71317 opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

renovate[bot] committed:
chore(deps): add renovate.json

glo71317 committed:
Added PHP 8.2 support, dropped PHP 7 support

Ocramius commented on .laminas-ci.json:
```diff
@@ -1,5 +1,5 @@
 {
     "ignore_php_platform_requirements": {
-        "8.1": true
+        "8.2": true
```
Probably not needed anymore? 🤔

glo71317 replied:
laminas/laminas-captcha has php8.2 support as this required php8.2 support

Ocramius replied:
Yes, but are current dependencies precluding a PHP 8 upgrade? This file introduces a CI error suppression/workaround, so it should be possible to remove it, if all dependencies are 8.2-ready

glo71317 replied:
Yes all dependencies have 8.2 support

Ocramius replied:
Then this file can be dropped 👍

glo71317 replied:
👍

Ocramius commented on composer.json:
```diff
@@ -21,7 +21,7 @@
     },
     "require": {
-        "php": "^7.3 || ~8.0.0 || ~8.1.0",
+        "php": "^7.3 || ~8.0.0 || ~8.1.0 || ~8.2.0",
```
We should drop PHP 7

Ocramius replied:
`config.platform` needs to pin the PHP version

glo71317 replied:
done

Ocramius commented:
Hmm, CI failures highlights some needs for changes.

glo71317 committed:
drop the laminas-ci.json file as all dependent packages has php8.2 support

Ocramius commented:
The test failures need to be addressed, not worked around 😁

glo82145 committed:
added  convertDeprecationsToExceptions in phpunit.xml.dist

Ocramius commented:
I think the minimum PHPUnit version needs to be raised to 9.5 here

glo82145 committed:
upgraded phpunit version to 9.5 in composer.json

glo82145 commented:
> I think the minimum PHPUnit version needs to be raised to 9.5 here

@Ocramius Thank you so much , this solution has worked here. Still One test case is failing, looking into that

glo82145 commented:
@Ocramius there is one testcase failing due to squizlabs/php_codesniffer. It could be due to same reason as you mentioned in PR https://github.com/laminas/laminas-oauth/pull/20. Can you please confirm?

Ocramius commented:
I wonder if we should just skip PHPCS: we don't want to improve CS on these components, as it poses more risk (regressions).

Alternatively, upgrading of `laminas/laminas-coding-standard` is possible, but with very conservative changes.

glo82145 commented:
> I wonder if we should just skip PHPCS: we don't want to improve CS on these components, as it poses more risk (regressions).
> 
> Alternatively, upgrading of `laminas/laminas-coding-standard` is possible, but with very conservative changes.

@Ocramius Yes, upgrading `laminas/laminas-coding-standard` would ask for lot of changes which will make things more complicated at this stage. Hence it should be investigated further to find appropriate solution.  But as it is blocking this current PR from merging and creating a blocker, Can we just skip this one  from this PR and take it up as different ticket for further investigation..

Ocramius commented:
> But as it is blocking this current PR from merging and creating a blocker

There is no blocker over here? Simply means it won't be merged until green here. 

 > Can we just skip this one from this PR and take it up as different ticket for further investigation..

IMO no: merging broken stuff is just going to lead to more load on the maintainers: we don't leave stuff half-done here, as it's mostly making thing complicated for everyone else.

Ocramius commented on phpcs.xml:
```diff
@@ -17,24 +17,4 @@
     <exclude-pattern>*/TestAsset/*</exclude-pattern>
-
-    <!-- Include all rules from Laminas Coding Standard -->
-    <rule ref="LaminasCodingStandard">
```
The `LaminasCodingStandard` should still be preserved

glo82145 commented:
@Ocramius We have upgraded the laminas/laminas-coding-standard and updated the code as per the requirement.

Ocramius committed:
Exclude `declare(strict_types=1)` enforcement from CS rules

Strict types are too problematic to enable in a minor release: we
will therefore disable the CS rule that enforces them to be declared.

This change also reverts the previous addition of strict types to
sources that can crash due to types mismatching internally.

Ocramius committed:
Included psalm static analysis to prevent minor regressions in the codebase

This codebase is tested, but is old and full of quirks: including static analysis
tooling will prevent minor changes from accidentally breaking it over minor details.

Ocramius committed:
Merge remote-tracking branch 'origin/renovate/configure' into php8.2-support

Ocramius committed:
Merge branch 'feature/add-psalm-static-analysis' into php8.2-support

Ocramius committed:
Removing unused test autoloader: was used for test classes that were part of PHPUnit < 9

Ocramius committed:
Exclude build artifacts from distribution package

Ocramius committed:
Adjusted `DecoratorManagerTest`

Invalid types were provied/documented in #12, so this
is a rewrite of those adjustments.

Ocramius committed:
Restored internal docs of `Table\Table`, broken by #12 work

Ocramius committed:
Restored internal docs of `Table\Row`, broken by #12 work

Ocramius committed:
Replaced `gettype` with `get_debug_type` when looking at unknown `mixed` value

Ocramius committed:
Adjusted `DecoratorManager` to adjust to `AbstractPluginManager` types

Also, this provides template parameters for better downstream type inference.

Ocramius committed:
Restored `Table\Column` documentation broken by #12

Ocramius committed:
Restored `Multibyte` documentation and deprecations broken by #12

Ocramius committed:
Restored `Figlet` documentation broken by #12

Ocramius committed:
Regenerated psalm baseline based on type improvements coming from the PHP and PHPCS upgrades

Ocramius committed:
Applied automated CS fixes: sorting of docblock annotation groups

Ocramius reviewed (approved):
I had to rewrite the patch to fix multiple changes done incorrectly, such as removed `@throws` descriptions, incorrect/wrong `@var` annotations and removed `@deprecated` comments.

I also squashed previous commits, which were just saying mentioning that some issues were being fixed, without any mention if which they were.

In practice, I understand that you were assigned this task (outsourced?), and that you had no experience with this, but I really wish you had put some thinking into the errors that the CS tooling reported, rather than just steamrolling over the pre-existing code.

Instead, it seems like you've bashed your keyboard until you reached green, which is a bit sad, since it just means double work for maintainers :-( 

Anyway, this is now ready for release.
